### DISCUSSION TITLE
Keep DAC ON during Standby Sleep Mode

### DISCUSSION THREAD

**Discussion by awc105** [SpenceKonde/megaTinyCore#885] [Q&A]
Hi,

I have a situation where I need to keep the DAC ON while the CPU is in standby sleep, how can I accomplish this?
I am using an ATTINY816 or ATTINY1616 @ 2.5V @ 5MHz 

For context I am using an external interrupt on PIN PC2 from an accelerometer to wake the CPU and change the DAC value, then go back to sleep. The problem is when the system goes to sleep, the DAC output also gets disabled.

Thanks,
Andrew

**Comment by hmeijdam** [comment]
See datasheet. It needs the peripheral clock to be active.

![image](https://user-images.githubusercontent.com/9595465/210310726-4efffca8-bfea-4dac-8c69-056cf71eeff1.png)

**Comment by awc105** [comment]
Hey Thanks for the tip!

I solved it using:
DAC0.CTRLA = 0xC1; //DAC0 Standby Enable, DAC0 Output Enable, DAC0 Enable ---> CTRLA = 11000001, page 458

and

CLKCTRL.OSC20MCTRLA = 0x02; //Clock_Per to run in standby mode ---> CTRLA = 00000010, page 85

**Comment by SpenceKonde** [comment]
It can't be run in power down sleep. But according to the datasheet, powerdown sleep on a 1-series tiny isn't necessarily any lower power than standby sleep - power down and standby sleep with everything disabled is listed in the same line of table 36.5 (based on the 3216 datasheet - you don't inherently get anything for going all the way to power down sleep, assuming you don't have any other peripherals wedged on with the RUNSTBY bit. So you need to be using SLEEP_MODE_STANDBY not SLEEP_MODE_POWER_DOWN, *and* you need to have set the run-standby bit.

```
DAC0.CTRLA |= DAC_RUNSTBY_bm; 
```

That said, it's a DAC, and DAC power consumption is 

With current released  version, if using analogWrite() to control the DAC, you need to do that after every call to analogWrite(), or just  manually configure the DAC (by setting DAC0.DATA=(8-bit value) and DAC0.CTRLA = 0xC1 (if manually controlling the DAC, you only need to change DAC0.DATA to change the value once CTRLA is set up (it looks like you're doing it manually

In analogWrite() I used simple assignment to write 0x41 (0x01 is enable, 0x40 is output buffer enable) to CTRLA; (the code does DAC0.DATA=val; DAC0.CTRLA=0x41; which compiles to sts 0x0680, r26; ldi r26, 0x41; sts 0x0681, r26; - 5 words 5 clocks)
I don't think current treatment of that is ideal - it ought not to unset runstby if the user has changed it. It costs 2 words of flash to do it the right way (DAC0.DATA=val; DAC0.CTRLA |= 0x41; compiles to sts 0x0680, r26; lds r26, 0x0681; ori r26, 0x41; sts 0x0681, r26; - 7 words and 8 clocks). (Could be hand-implemented if the z register is unused. Since z is call clobbered, and the sequence is immediately followed by a return, unless the compiler inlines it (which it won't do if the function is called from more than one place, it will be unused at that point in execution.  and would result in ldi ldi st ldd ori std, which would could do it in 6 words and 7 clocks, but I'm not going to write inline asm for just 2 bytes of flash - I try to refrain from throwing inline assembly around for single words, because it has a negative impact on code readability, since far more people know embedded-C than AVR ASM)

I've changed that so that analogWrite() won't forget the current runstby setting, and it will be in the next release. I *think* that LTO is clever enough that as long as you don't analogWrite() the DAC pin or a compile-time-unknown pin, it won't include that case in the generated code to prevent it from taking an extra 4 bytes of flash (though it will still forget the setting when you turn the pin off with digitalWrite(), which I would argue is correct behavior). 

Note: It is recommended that you set `PORTA.PIN6CTRL = PORT_ISC_INPUT_DISABLE_gc;` or `pinConfigure(PIN_PA6, PIN_INPUT_DISABLE);` when using the DAC if you care about power consumption. We don't automatically set that; it would add a dozen bytes to the code size, since we'd spend 3 words turning it on in analogWrite() and 3 words to turning it off in digitalWrite() (well, turnOffPWM()), and half of that burden would be applied to users who only use the pin for digital writes, making it three times as expensive as this change - for a convenience that in the only applicable use cases would be redundant with ordinary pin initialization best practices. If you care about power consumption, you need to either to either disable input buffers, enable the pullup, or set output all pins not connected to external digital levels. (see PowerSave.md)
